File: Codigo/2 - Ad-Hoc/src/2191.cpp
/**
* @file 2191.cpp
* Implementação O(n) - Kadane`s Algorithm
*
* Created on 2024-08-14 at 13:32:34
* @author GabrielCampelo
*/

#include <bits/stdc++.h>
#define _ ios_base::sync_with_stdio(0);cin.tie(0);
#define endl '\n'
#define pb push_back
#define all(x) (x).begin(), (x).end()

using namespace std;

typedef long long ll;
typedef unsigned long long llu;

const int INF = 0x3f3f3f3f;
const ll LINF = 0x3f3f3f3f3f3f3f3fll;

int main() { _
    int n;
    int testes = 1;
    while (cin >> n && n != 0) {
        cout << "Teste " << testes << endl;
        testes++;

        int mx = 0;
        int curr = 0;
        int mx_j, mx_i;
        int i = 0;
        for (int j = 0; j < n; j++) {
            int x, y;
            cin >> x >> y;
            curr += x - y;

            if (curr > mx) {
                mx_j = j;
                mx_i = i;
                mx = curr;
            } else if (curr == mx && j - i > mx_j - mx_i) {
                mx_j = j;
                mx_i = i;
            }

            if (curr < 0) {
                curr = 0;
                i = j + 1;
            }
        }

        if (mx == 0) cout << "nenhum" << endl << endl;
        else cout << mx_i + 1 << ' ' << mx_j + 1 << endl << endl;
    }
    
    return 0;
};

File: Codigo/2 - Ad-Hoc/src/2191.py
"""
Created on 2024-08-13 at 15:16:53

O(n) - Kadane`s Algorithm

@author: GabrielCampelo
"""

def main():
    testes = 1
    while True:
        n = int(input())
        if n == 0:
            break

        print(f"Teste {testes}")
        testes += 1

        saldo = [0] * n
        for i in range(n):
            x, y = map(int, input().split())
            saldo[i] = x - y

        # Implementação de Kadane para encontrar a maior soma de subarray
        max_sum = 0
        current_sum = 0
        start = 0
        end = 0
        temp_start = 0

        for i in range(n):
            current_sum += saldo[i]

            if current_sum > max_sum:
                max_sum = current_sum
                start = temp_start
                end = i
            elif current_sum == max_sum:
                if i - temp_start > end - start:
                    start = temp_start
                    end = i

            if current_sum < 0:
                current_sum = 0
                temp_start = i + 1

        if max_sum == 0:
            print("nenhum\n")
        else:
            print(f"{start + 1} {end + 1}\n")

if __name__ == '__main__':
    main()
